File: doc/Programs/LecturePrograms/programs/RandomWalks/Fortran/program2.f90
!  Program to perform a random walk in 2 dimensions
!  with equal probability for moves to the letf, right, up and down
!  Only one walker present

PROGRAM random_walk_2dim
  USE constants
  USE f90library
  IMPLICIT NONE
  INTEGER::  max_trials, number_walks
  REAL(dp):: move_probability
  INTEGER, ALLOCATABLE, DIMENSION(:) :: x_cum, x2_cum, y_cum, y2_cum

  ! get max trials, number of walks and the probability to move
  CALL  initialise(max_trials, number_walks, move_probability) 
  ! allocate memory
  ALLOCATE (x_cum(number_walks), x2_cum(number_walks), &
       y_cum(number_walks), y2_cum(number_walks))
  x_cum = 0; y_cum = 0; x2_cum = 0; y2_cum = 0  
  ! do the Monte Carlo move
  CALL  mc_sampling(max_trials, number_walks, move_probability, &
       x_cum, x2_cum, y_cum, y2_cum)
  ! Print out results 
  CALL  output(max_trials, number_walks, x_cum, x2_cum, y_cum, y2_cum) 
  ! free memory
  DEALLOCATE(x_cum, x2_cum, y_cum, y2_cum) 
END PROGRAM random_walk_2dim

! Read in data from screen

SUBROUTINE initialise(max_trials, number_walks, move_probability) 
  USE constants
  IMPLICIT NONE
  INTEGER, INTENT(inout) ::max_trials, number_walks
  REAL(DP) :: move_probability
  WRITE(*,*)'Number of Monte Carlo trials =' 
  READ(*,*)  max_trials
  WRITE(*,*) 'Number of attempted walks='
  READ(*,*) number_walks
  WRITE(*,*) 'Move probability='
  READ(*,*)  move_probability
END SUBROUTINE initialise

! Output of expectation values, <x>, <y> and their variances
! <x^2> - <x>^2  and <y^2> - <y>^2

SUBROUTINE output(max_trials, number_walks, x_cum, x2_cum, y_cum, y2_cum)
  USE constants
  IMPLICIT NONE
  INTEGER :: i
  INTEGER, INTENT(in) :: max_trials, number_walks
  INTEGER, INTENT(in) :: x_cum(number_walks), x2_cum(number_walks), &
       y_cum(number_walks), y2_cum(number_walks)
  REAL(dp) :: xaverage, yaverage , x2average, y2average, total_average, &
       total_variance, yvariance, xvariance

  OPEN(UNIT=6, FILE='out.dat')
  DO i =1, number_walks 
     xaverage = x_cum(i)/FLOAT(max_trials)
     x2average = x2_cum(i)/FLOAT(max_trials)
     xvariance = x2average - xaverage*xaverage
     yaverage = y_cum(i)/FLOAT(max_trials)
     y2average = y2_cum(i)/FLOAT(max_trials)
     yvariance = y2average - yaverage*yaverage
     total_average = xaverage+yaverage
     total_variance = xvariance+yvariance
     WRITE(6,'(I3,2X,6F12.6)') i, xaverage, yaverage ,xvariance, yvariance, &
          total_average, total_variance
  ENDDO
END SUBROUTINE output

! Here we perform the Monte Carlo samplings and moves in 2 dim

SUBROUTINE mc_sampling(max_trials,number_walks, move_probability, x_cum, &
     x2_cum, y_cum, y2_cum)
  USE constants
  USE f90library
  IMPLICIT NONE
  INTEGER, INTENT(in) :: max_trials,number_walks
  INTEGER, INTENT(inout) :: x_cum(number_walks), x2_cum(number_walks), &
       y_cum(number_walks), y2_cum(number_walks)
  INTEGER :: idum, x, y, trial, walks
  REAL(dp) :: rantest
  REAL(dp), INTENT(in) :: move_probability
  idum=-1 
  ! loop over Monte Carlo trials
  DO trial = 1, max_trials
     x = 0;  y = 0
     ! loop over attempted walks in 2 dimensions
     DO walks = 1, number_walks
        !  note code taylored to equal probability for up, down, left and right
        rantest = ran0(idum)
        IF (rantest <= move_probability) THEN
           x = x+1
        ELSEIF(rantest <= 2*move_probability)THEN
           x = x-1
        ELSEIF(rantest <= 3*move_probability)THEN
           y = y+1
        ELSE
           y = y-1
        ENDIF
        x_cum(walks) = x_cum(walks)+x
        x2_cum(walks) = x2_cum(walks)+x*x
        y_cum(walks) = y_cum(walks)+y
        y2_cum(walks) = y2_cum(walks)+y*y
     ENDDO
  ENDDO
END SUBROUTINE mc_sampling

File: doc/Programs/LecturePrograms/programs/RandomWalks/cpp/program2.cpp

/*
  Two dimensional random walk program. 
  A walker makes several trials steps with
  a given number of walks per trial
*/

#include <cmath>
#include <iostream>
#include <fstream>
#include <iomanip>
#include "lib.h"
using namespace  std;

// Function to read in data from screen, note call by reference
void initialise(int&, int&, double&) ;
// The Mc sampling for random walks
void  mc_sampling(int, int, double, int *, int *, int *, int *);
// prints to screen the results of the calculations 
void  output(int, int, int *, int *, int *, int *);

int main()
{
  int max_trials, number_walks; 
  double move_probability;
  // Read in data 
  initialise(max_trials, number_walks, move_probability) ;
  int *x_cum = new int [number_walks+1];
  int *x2_cum = new int [number_walks+1];
  int *y_cum = new int [number_walks+1];
  int *y2_cum = new int [number_walks+1];
  for (int walks = 1; walks <= number_walks; walks++){   
    x_cum[walks] = x2_cum[walks] = y_cum[walks] = y2_cum[walks] = 0;
  } // end initialization of vectors
  // Do the mc sampling  
  mc_sampling(max_trials, number_walks, move_probability, 
              x_cum, x2_cum, y_cum, y2_cum);
  // Print out results 
  output(max_trials, number_walks, x_cum, x2_cum, y_cum, y2_cum); 
  delete [] x_cum; // free memory
  delete [] x2_cum; delete [] y_cum; delete [] y2_cum;
  return 0; 
} // end main function

// note call by reference, differs from C++
void initialise(int& max_trials, int& number_walks, double& move_probability) 
{
  cout << "Number of Monte Carlo trials ="; 
  cin >> max_trials;
  cout << "Number of attempted walks=";
  cin >> number_walks;
  cout << "Move probability=";
  cin >> move_probability;
}  // end of function initialise   


void output(int max_trials, int number_walks, 
            int *x_cum, int *x2_cum, int *y_cum, int *y2_cum)
{
  ofstream ofile("twodim_walkers.dat");
  for( int  i = 1; i <=  number_walks; i++){
    double xaverage = x_cum[i]/((double) max_trials);
    double x2average = x2_cum[i]/((double) max_trials);
    double xvariance = x2average - xaverage*xaverage;
    double yaverage = y_cum[i]/((double) max_trials);
    double y2average = y2_cum[i]/((double) max_trials);
    double yvariance = y2average - yaverage*yaverage;
    double total_average = xaverage+yaverage;
    double total_variance = xvariance+yvariance;
    ofile << setiosflags(ios::showpoint | ios::uppercase);
    ofile << setw(6) << i;
    ofile << setw(15) << setprecision(8) << xaverage;
    ofile << setw(15) << setprecision(8) << xvariance; 
    ofile << setw(15) << setprecision(8) << yaverage;
    ofile << setw(15) << setprecision(8) << yvariance; 
    ofile << setw(15) << setprecision(8) << total_average;
    ofile << setw(15) << setprecision(8) << total_variance << endl;
  }
  ofile.close();

}  // end of function output 


void mc_sampling(int max_trials, int number_walks, 
                 double move_probability, int *x_cum, 
                 int *x2_cum, int *y_cum, int *y2_cum)
{
  long idum;
  idum=-1;  // initialise random number generator
  for (int trial=1; trial <= max_trials; trial++){
    int x = 0; int y = 0;
    for (int walks = 1; walks <= number_walks; walks++){   
      double rantest = ran0(&idum);
      if (rantest <= move_probability) {
	x += 1;
      } 
      else if (rantest <= 2*move_probability){
	x -= 1;
      }
      else if (rantest <= 3*move_probability){
        y += 1;
      }
      else {
        y -= 1;
      }
      x_cum[walks] += x;
      x2_cum[walks] += x*x;
      y_cum[walks] += y;
      y2_cum[walks] += y*y;
    }  // end of loop over walks
  } // end of loop over trials
}   // end mc_sampling function  




